snakemake/executors.py

```diff
@@ -2168,3 +2171,878 @@ def run_wrapper(
         except (Exception, BaseException) as ex:
             raise WorkflowError(ex)
+
+
+GoogleLifeSciencesJob = namedtuple(
+    "GoogleLifeSciencesJob", "job jobname jobid callback error_callback"
+)
+
+
+class GoogleLifeSciencesExecutor(ClusterExecutor):
+    """the GoogleLifeSciences executor uses Google Cloud Storage, and
+       Compute Engine paired with the Google Life Sciences API.
+       https://cloud.google.com/life-sciences/docs/quickstart
+    """
+
+    def __init__(
+        self,
+        workflow,
+        dag,
+        cores,
+        envvars,
+        jobname="snakejob.{name}.{jobid}.sh",
+        printreason=False,
+        quiet=False,
+        printshellcmds=False,
+        container_image=None,
+        regions=None,
+        location=None,
+        cache=False,
+        machine_type_prefix=None,
+        latency_wait=3,
+        local_input=None,
+        restart_times=None,
+        exec_job=None,
+        max_status_checks_per_second=1,
+    ):
+
+        # Attach variables for easy access
+        self.workflow = workflow
+        self.quiet = quiet
+        self.workdir = os.path.dirname(self.workflow.persistence.path)
+        self._save_storage_cache = cache
+        self._machine_type_prefix = machine_type_prefix
+
+        # Relative path for running on instance
+        self._set_snakefile()
+
+        # Prepare workflow sources for build package
+        self._set_workflow_sources()
+
+        exec_job = exec_job or (
+            "snakemake {target} --snakefile %s "
+            "--force -j{cores} --keep-target-files --keep-remote "
+            "--latency-wait 0 "
+            "--attempt 1 {use_threads} "
+            "{overwrite_config} {rules} --nocolor "
+            "--notemp --no-hooks --nolock " % self.snakefile
+        )
+
+        # IMPORTANT: using Compute Engine API and not k8s == no support secrets
+        self.envvars = envvars or []
+
+        # Quit early if we can't authenticate
+        self._get_services()
+        self._get_bucket()
+
+        # Akin to Kubernetes, create a run namespace, default container image
+        self.run_namespace = str(uuid.uuid4())
+        self.container_image = container_image or get_container_image()
+        self.regions = regions or ["us-east1", "us-west1", "us-central1"]
+
+        # Determine API location based on user preference, and then regions
+        self._set_location(location)
+
+        # Tell the user right away the regions, location, and container
+        logger.debug("regions=%s" % self.regions)
+        logger.debug("location=%s" % self.location)
+        logger.debug("container=%s" % self.container_image)
+
+        # The project name is required, either from client or environment
+        self.project = (
+            os.environ.get("GOOGLE_CLOUD_PROJECT") or self._bucket_service.project
+        )
+        if not self.project:
+            raise WorkflowError(
+                "You must provide a --google-project or export "
```

Comment:
Adapt the error message, --google-project is not available anymore.

Reply:
I think I can remove this entirely because `self._bucket_service.project` is always defined, granted that we successfully grab a bucket!

Reply:
I can't comment on the one above, but this comment:

> Ok, now I understand what you do here. You can instead just use os.getcwd()

That won't work because the workdir is definitely not always (usually not) the os.getcwd() Can you explain why this would work? (e.g., when we run a test and move to a temporary location, the working directory is the temporary location but os.getcwd() is still my present working directory where I'm running from).